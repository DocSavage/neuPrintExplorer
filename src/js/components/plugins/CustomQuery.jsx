/*
 * Supports simple, custom neo4j query.
*/
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import randomColor from 'randomcolor';
import { withRouter } from 'react-router';

import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';

import { submit } from 'actions/plugins';
import { getQueryString } from 'helpers/queryString';
import { LoadQueryString, SaveQueryString } from '../../helpers/qsparser';
import { setUrlQS } from '../../actions/app';

const styles = () => ({
  textField: {
    width: 300
  },
  button: {
    margin: 4,
    display: 'block'
  },
  formControl: {}
});

const pluginName = 'CustomQuery';

class CustomQuery extends React.Component {
  static get queryName() {
    return 'Custom';
  }

  static get queryDescription() {
    return 'Enter custom Neo4j Cypher query';
  }

  constructor(props) {
    super(props);
    var initqsParams = {
      textValue: ''
    };
    var qsParams = LoadQueryString(
      'Query:' + this.constructor.queryName,
      initqsParams,
      this.props.urlQueryString
    );
    this.state = {
      qsParams: qsParams
    };
  }

  processResults = (query, apiResponse) => {
    if (apiResponse.data) {
      return {
        columns: apiResponse.columns,
        data: apiResponse.data,
        debug: apiResponse.debug
      };
    } else {
      return {
        columns: [],
        data: [],
        debug: ''
      };
    }
  };

  processRequest = () => {
    const { dataSet, actions, history } = this.props;
    const { textValue } = this.state.qsParams;

    const query = {
      dataSet,
      cypherQuery: textValue,
      visType: 'SimpleTable',
      plugin: pluginName,
      parameters: {},
      title: 'Custom query',
      menuColor: randomColor({ luminosity: 'light', hue: 'random' }),
      processResults: this.processResults
    };
    actions.submit(query);
    history.push({
      pathname: '/results',
      search: getQueryString()
    });
    return query;
  };

  handleChange = event => {
    const oldParams = this.state.qsParams;
    oldParams.textValue = event.target.value;
    this.props.actions.setURLQs(SaveQueryString('Query:' + this.constructor.queryName, oldParams));
    this.setState({
      qsParams: oldParams
    });
  };

  catchReturn = event => {
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      this.processRequest();
    }
  };

  render() {
    const { classes } = this.props;
    return (
      <FormControl className={classes.formControl}>
        <TextField
          label="Custom Cypher Query"
          multiline
          value={this.state.qsParams.textValue}
          rows={1}
          rowsMax={4}
          className={classes.textField}
          onChange={this.handleChange}
          onKeyDown={this.catchReturn}
        />
        <Button
          variant="contained"
          className={classes.button}
          onClick={this.processRequest}
          disabled={!(this.state.qsParams.textValue.length > 0)}
          color="primary"
        >
          Submit
        </Button>
      </FormControl>
    );
  }
}

CustomQuery.propTypes = {
  urlQueryString: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired
};

var CustomQueryState = function(state) {
  return {
    urlQueryString: state.app.get('urlQueryString'),
    isQuerying: state.query.isQuerying
  };
};

var CustomQueryDispatch = dispatch => ({
  actions: {
    submit: query => {
      dispatch(submit(query));
    },
    setURLQs: function(querystring) {
      dispatch(setUrlQS(querystring));
    }
  }
});

export default withStyles(styles)(
  withRouter(
    connect(
      CustomQueryState,
      CustomQueryDispatch
    )(CustomQuery)
  )
);
