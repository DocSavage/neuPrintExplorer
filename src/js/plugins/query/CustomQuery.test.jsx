import React from 'react';
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-unresolved
import '@testing-library/jest-dom';
import 'regenerator-runtime';
import { CustomQuery } from './CustomQuery';

const { submit } = global;

const styles = { textField: '', button: '', formControl: '' };

const component = (
  <CustomQuery
    dataSet="test"
    history={{ push: jest.fn() }}
    classes={styles}
    submit={submit}
    isQuerying={false}
    cypherFromOpenTab="test"
  />
);

describe('custom query Plugin', () => {
  beforeEach(() => {
    submit.mockClear();
  });
  it('has name and description', () => {
    expect(CustomQuery.details.name).toBeTruthy();
    expect(CustomQuery.details.description).toBeTruthy();
  });
  it('renders correctly', () => {
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });
  /*
  describe('when user clicks submit', () => {
    it('should return a query object and submit', () => {
      expect(button.props().onClick()).toEqual(undefined);
      expect(submit).toHaveBeenCalledTimes(1);
    });

    it('should process returned results into data object', () => {
      const query = {
        dataSet: 'test',
        cypherQuery: 'test',
        visType: 'SimpleTable',
        plugin: 'CustomQuery',
        parameters: {},
        title: 'Custom query'
      };
      const apiResponse = {
        data: [[1, 2, 3], [4, 5, 6]],
        columns: ['a', 'b', 'c'],
        debug: 'test',
        title: 'Custom Query'
      };
      const processedResults = CustomQuery.processResults(query, apiResponse);
      expect(processedResults).toEqual(apiResponse);

      // if no data returned
      const processedResultsEmpty = CustomQuery.processResults(query, {});
      expect(processedResultsEmpty).toEqual({
        columns: [],
        data: [],
        debug: ''
      });
    });
  });

  describe('when user hits enter key', () => {
    test('should submit request', async () => {
      render(component);
      screen.debug();
      const textField = await screen.getByRole('textbox');
      await userEvent.type(textField, '{enter}');
      expect(submit).toHaveBeenCalledTimes(1);
    });
  });

  describe('when user inputs text', () => {
    it('should change state', () => {
      actions.setQueryString.mockClear();
      textField.props().onChange({ target: { value: 'abc' } });
      expect(wrapper.find('CustomQuery').state('textValue')).toEqual('abc');
    });
  }); */
});
