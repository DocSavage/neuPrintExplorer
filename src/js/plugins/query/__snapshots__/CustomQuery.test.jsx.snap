// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom query Plugin renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth"
  >
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Custom Cypher Query
    </p>
    <div
      class="react-codemirror2"
    >
      <div
        class="CodeMirror cm-s-default CodeMirror-wrap"
        translate="no"
      >
        <div
          style="overflow: hidden; position: relative; width: 3px; height: 0px;"
        >
          <textarea
            autocapitalize="off"
            autocorrect="off"
            spellcheck="false"
            style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; min-height: 1em; outline: none;"
            tabindex="0"
          />
        </div>
        <div
          class="CodeMirror-vscrollbar"
          cm-not-content="true"
          tabindex="-1"
        >
          <div
            style="min-width: 1px;"
          />
        </div>
        <div
          class="CodeMirror-hscrollbar"
          cm-not-content="true"
          tabindex="-1"
        >
          <div
            style="height: 100%; min-height: 1px;"
          />
        </div>
        <div
          class="CodeMirror-scrollbar-filler"
          cm-not-content="true"
        />
        <div
          class="CodeMirror-gutter-filler"
          cm-not-content="true"
        />
        <div
          class="CodeMirror-scroll"
          tabindex="-1"
        >
          <div
            class="CodeMirror-sizer"
            style="margin-left: 0px;"
          >
            <div
              style="position: relative;"
            >
              <div
                class="CodeMirror-lines"
                role="presentation"
              >
                <div
                  role="presentation"
                  style="position: relative; outline: none;"
                >
                  <div
                    class="CodeMirror-measure"
                  >
                    <pre
                      class="CodeMirror-line-like"
                    >
                      <span>
                        xxxxxxxxxx
                      </span>
                    </pre>
                    <div
                      class="CodeMirror-linenumber CodeMirror-gutter-elt"
                    >
                      <div>
                        1
                      </div>
                    </div>
                  </div>
                  <div
                    class="CodeMirror-measure"
                  />
                  <div
                    style="position: relative; z-index: 1;"
                  />
                  <div
                    class="CodeMirror-cursors"
                  />
                  <div
                    class="CodeMirror-code"
                    role="presentation"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            style="position: absolute; height: 50px; width: 1px;"
          />
          <div
            class="CodeMirror-gutters"
          >
            <div
              class="CodeMirror-gutter CodeMirror-linenumbers"
              style="width: 1px;"
            />
          </div>
        </div>
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption"
    >
      [shift] + [Enter] to submit.
    </span>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
